abci/client/committing_client.go

```diff
@@ -22,17 +24,52 @@ type committingClient struct {
 	// Commit
 	// ApplySnapshotChunk
 	mtx *tmsync.RWMutex
+
+	// If called with false, we will wait until the app is ready to serve queries
+	// and deliveries.
+	// CONTRACT: The mtx read lock must be held.
+	//
+	// If called with true, we will indicate that the app is ready.
+	// CONTRACT: The mtx write lock must be held.
+	guardUntilStateIsCommitted func(bool)
+
 	types.Application
 	Callback
 }
 
-func NewCommittingClient(mtx *tmsync.RWMutex, app types.Application) Client {
+// NewCommittingClientGuard uses a condition variable to implement a guard.
+func NewCommittingClientGuard(mtx *tmsync.RWMutex) func(bool) {
+	waitWhileGuardIsActive := sync.NewCond(mtx.RLocker())
```

Comment:
It took me a while to convince myself that this is legit. (The `sync` package docs say that it's okay to use a `RWMutex` as the mutex for a condition variable, but not necessarily via its read lock.) Please add the following comment to help future troubleshooters:

```
// Condition variables should work with read locks, as long as the actual state update is done
// under the associated write lock to provide happens-before visibility. See comment for
// notifyListAdd() in src/runtime/sema.go in the Go source code.
```

Reply:
Done.